pkg/storage/iterators.go

```diff
@@ -53,28 +55,36 @@ func (c *combinedIterator[T]) Next(ctx context.Context) (T, error) {
 		return val, ErrIteratorDone
 	}
 
+	c.mu.Lock()
 	iter := c.pending[0]
 	val, err := iter.Next(ctx)
 	if err != nil {
 		if errors.Is(err, ErrIteratorDone) {
 			c.pending = c.pending[1:]
 			c.done = append(c.done, iter)
+			c.mu.Unlock()
```

Comment:
The introduction of this mutex is necessary as `c.pending` and `c.done` get accessed via multiple goroutines and can cause data race issues (and there are other methods which ultimately access these properties such as `.Stop`, thus a mutex is required.